third_party/nvfuser/test/test_gpu3.cpp

```diff
@@ -7774,6 +7774,59 @@ TEST_F(NVFuserTest, FusionPredicateReductionInitGlobal_CUDA) {
       fe.kernel(), cg_outputs, inputs, {ref_t1, ref_t3}, __LINE__, __FILE__);
 }
 
+TEST_F(NVFuserTest, FusionExecutorCache_IndexType_Recompilation) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  auto tv0 = makeContigTensor(2, DataType::Half);
+  auto tv1 = makeContigTensor(2, DataType::Half);
+  auto tv2 = add(tv0, tv1);
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv2);
+
+  auto fec = std::make_unique<FusionExecutorCache>(std::move(fusion));
+
+  auto options = at::TensorOptions().dtype(at::kHalf).device(at::kCUDA, 0);
+  auto input_small = at::ones({1024, 1024}, options);
+  auto input_large = at::ones({1024, 4 * 1 << 20}, options);
+
+  KernelArgumentHolder small_holder =
+      KernelArgumentHolder::createKernelArgumentHolder({input_small});
+  TORCH_CHECK(
+      small_holder.getIndexMode() ==
+      KernelIndexMode::INT32); // Index mode should be INT32
+
+  small_holder.setIndexMode(
+      KernelIndexMode::INT64); // Index mode should go up in size.
+  TORCH_CHECK(small_holder.getIndexMode() == KernelIndexMode::INT64);
+
+  small_holder.push(input_large);
+  small_holder.setIndexMode(KernelIndexMode::INT32);
+  TORCH_CHECK(
+      small_holder.getIndexMode() ==
+      KernelIndexMode::INT64); // Index mode shouldn't be changed.
+
+  // Nothing should be compiled prior to running a first time
+  TORCH_CHECK(!fec->isCompiled({input_small, input_small}));
+  TORCH_CHECK(!fec->isCompiled({input_large, input_large}));
+
+  // Only INT32 Indexing version should be compiled
+  fec->runFusionWithInputs({input_small, input_small});
+  TORCH_CHECK(fec->isCompiled({input_small, input_small}));
+  TORCH_CHECK(!fec->isCompiled({input_large, input_large}));
+
+  // Running input that requires INT64 should not delete cache entry for INT32
+  fec->runFusionWithInputs({input_large, input_large});
+  TORCH_CHECK(fec->isCompiled({input_small, input_small}));
+  TORCH_CHECK(fec->isCompiled({input_large, input_large}));
+
+  // In the other order is should also not delete the entry
+  fec->runFusionWithInputs({input_small, input_small});
+  TORCH_CHECK(fec->isCompiled({input_small, input_small}));
+  TORCH_CHECK(fec->isCompiled({input_large, input_large}));
```

Comment:
This doesn't seem to guarantee no recompilation happened when the fusion was executed again with the smaller inputs. I *think* this would work:

```
// first run with the smaller input
fec->runFusionWithInputs({input_small, input_small});
auto kernel_runtime_for_input_small = fec->getMostRecentKernelRuntime();

// run with the larger input
fec->runFusionWithInputs({input_large, input_large});
auto kernel_runtime_for_input_large = fec->getMostRecentKernelRuntime();

// make sure they correspond to different runtimes, just in case
TORCH_CHECK(kernel_runtime_for_input_small != kernel_runtime_for_input_large, "Unexpected kernel runtime");

// second run with the smaller input
fec->runFusionWithInputs({input_small, input_small});
auto second_kernel_runtime_for_input_small_ = fec->getMostRecentKernelRuntime();

TORCH_CHECK((kernel_runtime_for_input_small == second_kernel_runtime_for_input_large, "Expected to reuse the same runtime");
```

@jjsjann123 , can you comment if this makes sense?

Reply:
@naoyam so should I go ahead?

Reply:
Let me take a quick look there. Sorry I missed this PR during the break :pleading_face:

Reply:
I was under the impression that we are still using int32 indexing when smaller inputs were encountered after a recompilation.

Naoya's change looks good to me, except the last line, I think it's a typo and you mean  `second_kernel_runtime_for_input_small` on the left of the equation.

Reply:
Thanks. Fixed the above code.

@mmigdal-nv, please update the test.

Reply:
Done. They pass!